src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -61,28 +62,38 @@ void ChunkPruningRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) co
   for (ChunkID chunk_id{0}; chunk_id < table->chunk_count(); ++chunk_id) {
     statistics.push_back(table->get_chunk(chunk_id)->statistics());
   }
-  std::set<ChunkID> excluded_chunk_ids;
+  std::set<ChunkID> pruned_chunk_ids;
   for (auto& predicate : predicate_nodes) {
-    auto new_exclusions = _compute_exclude_list(statistics, *predicate->predicate(), *stored_table);
-    excluded_chunk_ids.insert(new_exclusions.begin(), new_exclusions.end());
+    auto new_exclusions = _compute_exclude_list(statistics, *predicate->predicate(), stored_table);
+    pruned_chunk_ids.insert(new_exclusions.begin(), new_exclusions.end());
   }
 
-  // wanted side effect of usings sets: excluded_chunk_ids vector is sorted
-  auto& already_excluded_chunk_ids = stored_table->excluded_chunk_ids();
-  if (!already_excluded_chunk_ids.empty()) {
+  // wanted side effect of usings sets: pruned_chunk_ids vector is sorted
+  auto& already_pruned_chunk_ids = stored_table->pruned_chunk_ids();
+  if (!already_pruned_chunk_ids.empty()) {
     std::vector<ChunkID> intersection;
-    std::set_intersection(already_excluded_chunk_ids.begin(), already_excluded_chunk_ids.end(),
-                          excluded_chunk_ids.begin(), excluded_chunk_ids.end(), std::back_inserter(intersection));
-    stored_table->set_excluded_chunk_ids(intersection);
+    std::set_intersection(already_pruned_chunk_ids.begin(), already_pruned_chunk_ids.end(), pruned_chunk_ids.begin(),
+                          pruned_chunk_ids.end(), std::back_inserter(intersection));
+    stored_table->set_pruned_chunk_ids(intersection);
   } else {
-    stored_table->set_excluded_chunk_ids(std::vector<ChunkID>(excluded_chunk_ids.begin(), excluded_chunk_ids.end()));
+    stored_table->set_pruned_chunk_ids(std::vector<ChunkID>(pruned_chunk_ids.begin(), pruned_chunk_ids.end()));
   }
 }
 
 std::set<ChunkID> ChunkPruningRule::_compute_exclude_list(
     const std::vector<std::shared_ptr<ChunkStatistics>>& statistics, const AbstractExpression& predicate,
-    const StoredTableNode& stored_table_node) const {
-  const auto operator_predicates = OperatorScanPredicate::from_expression(predicate, stored_table_node);
+    const std::shared_ptr<StoredTableNode>& stored_table_node) const {
+  // `statistics` contains SegmentStatistics for all columns, even those that are pruned in `stored_table_node`.
+  // To be able to build a OperatorScanPredicate that contains a ColumnID referring to the correct SegmentStatistics in
+  // `statistics`, we create a clone of `stored_table_node` without the pruning info.
```

Comment:
It's terrible. There never seems to be a refactoring that is just smooth... :(

I see no clean way around this, unfortunately. 

I have marked the 4 lines as "hacky", though....

Reply:
Why are you talking about an OperatorScanPredicate?

Is this "hack" also done in other operations that remove columns (cannot think of one right now)?

Reply:
> Why are you talking about an OperatorScanPredicate?

One is created six lines below, that's why.  

> Is this "hack" also done in other operations that remove columns (cannot think of one right now)?

Nope. The reason this case is special is because there is no good place to prune the ChunkStatistics... Could be done in a potential `StoredTableNode::get_chunk_statistics(ChunkID)` but that requires the copying of a lot of shared_ptrs. Think 500 Chunks and 20 columns surviving the pruning -> 10k copies to build the pruned ChunkStatistics.

Reply:
Ah I see.